### ISSUE TITLE
[css-paint-api] Current Color in context

### ISSUE THREAD

**Issue by nucliweb** [w3c/css-houdini-drafts#921]
Hi!

I want to use Paint API to generate a random background-color in an element, but I would like to get the currentColor of context to generate a random color with contrast.

Can I access to the currentColor?

Thanks

**Comment by AmeliaBR** [comment]
If your paint module code specifically wants the current color (ie, to _always_ match current color), you can declare a dependency on the `color` property via the [`inputProperties` getter](https://drafts.css-houdini.org/css-paint-api/#paint-definition-input-properties).

However, it gets more complicated if your module is dependent on another color input (a parameter, custom property, or another property like `border-color`) & the website CSS uses `currentColor` as the input value.  Since `currentColor` is a valid computed value, that styleMap that gets passed to your paint code will only have the keyword, instead of the final used color.

If you try to use the keyword directly in your painting code, it's currently [specified to behave as solid black](https://drafts.css-houdini.org/css-paint-api/#paint-definition-input-properties). Which is surely not what you intended when you passed `currentColor` as a parameter to your paint function.

**Comment by nucliweb** [comment]
Exactly, what I want is to be able to obtain the computed color of currentColor, since it would be the inherited color. I do not want to have to force to indicate the color in the element where the Paint API is being used.

Thanks @AmeliaBR 

```Btw, I tried to access to `styleMap` in Chrome but I don't get anything.```

**Comment by AmeliaBR** [comment]
> Btw, I tried to access to `styleMap` in Chrome but I don't get anything.

The name `styleMap` is just an example from the specs, the name you use will depend on how you name the parameters in your `paint` function. In the examples later in the spec, they are named `properties`.  Either way, it is the third parameter passed to your `paint` function.

That allows you to get a [`CSSStyleValue`](https://drafts.css-houdini.org/css-typed-om-1/#cssstylevalue) object representing the color value. _But_, the Typed OM spec doesn't yet define a specific type for a CSSColorValue object, so you need to parse the string value yourself.

I put together a demonstration, to test out how difficult it was to do (parsing the color string is the worst part) & to see how an explicit `currentColor` value messes things up: 

- Code: https://glitch.com/~obsidian-suit
- Result: https://obsidian-suit.glitch.me/

Chrome, it turns out, still returns RGB values for `currentColor`, so there's nothing special there.  (This isn't Paint-specific, you can also see it if you call `computedStyleMap()` on the element from the main document.) But, if they matched the spec and returned `currentColor` as a CSS keyword value, then you wouldn't be able to do any color math whatsoever.

Thinking over what to do about this when we actually _do_ define `CSSColorValue` types:

`currentColor` is defined to be a keyword at computed time so that it _inherits_ as a keyword. But there isn't any reason that the actual color value can't be known at computed time.  Maybe it would be better to think of `currentColor` in Typed OM as a full color object (e.g., that represents a specific RGB color) that _also_ has a flag that says "but this is a `currentColor` special color" for inheritance and serialization.

Until we figure out the best way to handle CSS colors that inherit as keywords (which may one day include [other keywords](https://github.com/w3c/csswg-drafts/issues/3847)), **I think we should remove all the property-specific rules in [CSS Typed OM](https://drafts.css-houdini.org/css-typed-om-1/#reify-property) that require `currentColor` to be returned as a `CSSKeywordValue` object.**

There's clearly no web-compat obstacle to changing it at this point. And we may have web compat problems in the future if Chrome tries to follow the current spec, since that would break paint worklets that are using `currentColor` as a value to regular CSS properties that are then input into the worklet.

I don't know what an ideal Typed OM representation of `currentColor` (or `buttonFace`) will be, but I think it should be a subclass of a `CSSColorValue` object, not a generic keyword. (Also, if the decision is to handle it as a keyword, that shouldn't be dealt with via "property-specific rules", but by general rules for handling computed values of colors.)

**Comment by nucliweb** [comment]
Thanks a lot @AmeliaBR, for the extended explanation and for the example 😊

You're breathtaking 🤘🏼

**Comment by tabatkins** [comment]
Agenda+ to discuss this.

There's no reason we *couldn't* give the painter used-value colors; it's an endpoint and that's fine and appropriate. (It's how all the other built-in painters work.) 

However, we don't have the infrastructure set up thru the specs to *generally* give access to a UsedStylePropertyMap object; that requires a *lot* of design work to rigorously define used values in all circumstances.

We might be able to offer color-resolving abilities thru a TypedOM API; it just needs an element for context, so it can check 'color' and 'color-scheme'.

Worst case, we can offer a special-purpose API to the PaintWorklet that resolves colors for you.

**Comment by AmeliaBR** [comment]
We have two factors at this point to consider:

- Web compat with existing content built for the shipped Chrome implementation, that expect the `toString()` result of the [`CSSStyleValue`](https://drafts.css-houdini.org/css-typed-om/#cssstylevalue) object for color properties to always return an `rgb()` or `rgba()` function.  (Most such content wouldn't need to *parse* the function like my demo did, but they would need it to be valid when used directly to set a canvas API color property.)

- The long-term goals of (a) having `currentColor` inherit as a keyword on all browsers, and (b) having a useful `CSSColorValue` object that can represent all color syntaxes (keywords and functions) and maybe do useful things like converting between function types.

I think these constraints can be satisfied if we say that:

- `currentColor` is first and formost a _color_ value, rather than a generic keyword. In Typed OM, it will be represented by `CSSColorValue`, just like other color keywords will be, not by `CSSKeywordValue`.

- the future `CSSColorValue` type will have some way of associating it (maybe an internal slot, maybe exposed by the API) with an palette of resolvable values for `currentColor` and the system color keywords, and these values will automatically be filled in when the color value is part of a computed style.

- the stringifier for `CSSColorValue` should return resolved rgb/rgba values whenever possible, using the associated element palette if it exists.

> Worst case, we can offer a special-purpose API to the PaintWorklet that resolves colors for you.

This is an alternative. Or just allow the paint context APIs to take contextual color keywords and resolve them against the element being painted.
